### ISSUE TITLE
Document cache (1000) section

### ISSUE THREAD

**Issue by smithfred** [ghacksuserjs/ghacks-user.js#333]
tl;dr: proper summary documentation is actually necessary for the 1000 section, per #51.

There was discussion on #51 about how (much) to document disabling caching etc. in section 1000, and some explanatory text discussed there was removed (seemingly on the basis of "Everyone will know what Cache is and does, and whether they want it enabled or not.")

Problem is, although the user of ghacks-user.js might (and probably does) know plenty of standard stuff about browser caching, what they *don't* know, is if there's some disk cache-based privacy (or other) attack/risk they haven't read about.

I know in my case, when I came to this section, my immediate thought was "Wait, why is cache being disabled? Is this just an "on-disk browsing info available locally" problem? Is it a tracking issue (ETags?) Is there some other attack vector I don't know about they're preventing here?"

So seeing the mostly undocumented inclusion of cache-disabling items, they now *don't* know if they can safely override them.

Compare signon.autofillForms. If the user keeps their system locally secure, they might think that feature is harmless, whereas in that case, there is both a 3rd-party and 1st-party tracking issue with leaving it enabled.

Is there a similar problem with cache? Without 1000 docs, the user doesn't know.

**Comment by Thorin-Oakenpants** [comment]
I've just re-read all of issue 51, and I'm not sure what to do now. Is it really not that obvious?

The main reason for disabling cache is separation of app and OS as per [TBB's](https://www.torproject.org/projects/torbrowser/design/index.html.en) design model of `State Separation` and `Disk Avoidance`

This is riddled throughout the user.js - think forensics
- opening (a pdf) in an external editor can leave an MRU (most recently used) <- state separation
- windows jump lists <- state separation
- clearing histories (form/download/etc) on close <- disk avoidance
- resistFingerprinting & 4605 -> do not store site specific prefs for zoom views <- disk avoidance
- loads more examples

Effectively - limit and keep all data (including cloud) within the app, cleanup the app when not used/closed, and the degrees to which you can apply this vary (1006 puts sites prefs into memory, 1221 does the same for logins/passwords/CA's ). So you can totally go hard core (or even use a FF in a VM that doesn't save changes) - the beauty is that if you are caught with your device logged in, there is less "evidence" - you can even go cache memoryless. As for encrypting/securing devices, I refer to the rubber hose scenario, or a severed (optional) finger and/or detached (again optional, but preferred, on a ball point pen) eyeball etc for biometrics. So the less data written, the better << disk avoidance

There is also that thing about writes to SSDs

So... I do not think it's just something that could be added to the cache section as it's way more widespread than that. It's also not an easy short thing to sum up in a line

PS: In earlier versions of the js on ghacks, I used to add things like `- FORENSICS` and `- SHOULDER SURFERS` at the end of numbered lines but at some point it becomes ridiculous and silly looking (well at least to me, esp as the entries grew)

**Comment by smithfred** [comment]
Users of ghacks-user.js will have different goals/non-goals. For instance:

* Prevent tracking/fingerprinting
* Prevent data-sharing with third parties (e.g. Google)
* Mitigate/prevent data breaches/exfiltration (e.g. history extraction)
* Mitigate/prevent system compromise attack vectors
* Prevent shoulder-surfing (+public space CCTV etc.) data gathering
* Prevent local storage of browsing history etc.

Personally my main purpose is "my personal life/interests are none of $(advertising/tracking/behemoth corp)'s fucking business", with a side of "arbitrary JS is a security trash-fire waiting to happen", but "I can keep my own machines sufficiently secure" ;)

Since the ghacks-user.js project seems to be characterised as covering different needs (customise per user's needs), then this info is needed somewhere. If not in the actual user.js file, then in an external file referenced by user.js (or easily cross-referenced).

I'm not sure that "silly-looking" or otherwise is really an important consideration, unless it actually affects usability.

A compact option might be an abbreviated key, e.g. /* F, SS - see user.js-notes.txt#1050 */

I filed this against the 1000 section since it was the first one where I was really unclear about the ramifications (a lot of others have a short explanation, or were self-explanatory to me), but you may well be right that other under-documented sections could benefit from the same.

**Comment by Thorin-Oakenpants** [comment]
I do not think we need to tag everything with `DA`, `F`, `SS` (or however we do it) - that is a lot of work. But thinking thru a list of what this js does (and there is a lot of cross-over and it would get "complicated" or messy rather fast)

- Disk Avoidance - don't put it there in the first place - TBB does this
- Persistent Data - clean it up on close - a balance to Disk Avoidance (eg we allow history but clear it on close)
- Shoulder Surfers - need a better word/phrase
- Security
- Privacy

There's a bit of an intro in the 0800 section, and in fact that section still contains those `- PRIVACY` etc line endings.

> but you may well be right that other under-documented sections could benefit from the same.

I think we need to do two things
- Create a simple list of what types of threats this user js tackles (see above list)
- Decide which sections and prefs of the user.js need some clarification

Once we have those two things, then we know the scope of what is required

**Comment by smithfred** [comment]
OK, I nominate the cache section for the second thing ;)

BTW great work on the whole project, much appreciated, and becoming ever more useful/necessary given Mozilla's current "leadership" and "decision-making" :J ...!

**Comment by Thorin-Oakenpants** [comment]
lol .. I already gathered that

**Comment by Thorin-Oakenpants** [comment]
I guess disk cache is faster than memory cache? - https://www.reddit.com/r/firefox/comments/7pld2l/back_navigation_slow/

I don't follow that - a backpage would be recent and surely RAM is quicker than reading the disk? IDK

---
Anyway: I am not going to write a manifesto or design document. Not going to label everything. Not going to bloat the readme section. Do not really wish to create a wiki page with stuff on it to maintain (and info is better off in the actual user.js IMO). As much as I would like to write something short and simple as to **general** rules (eg Disk Avoidance, Application Data Isolation, Lowering Entropy etc etc as per TBB design document), it's just too much IMO.

As for cache, I will think of something to put in the header section

**Comment by smithfred** [comment]
There's not actually extra work in documenting this somewhere (other than in GH issue discussions), since... well, it's already being documented already, in GH issue discussions...

> PS: In earlier versions of the js on ghacks, I used to add things like - FORENSICS and - SHOULDER SURFERS at the end of numbered lines but at some point it becomes ridiculous and silly looking

> and info is better off in the actual user.js IMO

Pick one, you can't have both ;)

Why not just have a user.js and a user.js-notes.md, and split the docs out to the latter entirely? Keeps the user.js clean, and lets you semantically mark up the notes (for notes navigation etc.)

**Comment by Thorin-Oakenpants** [comment]
I like all the info in the one file so it's easier for end users, but sure, valid points/ideas. I'll come back to it tomorrow - I need to think and sleep on it

> Pick one, you can't have both ;)

I can do whatever I like. I will have three of each, maybe.

**Comment by Thorin-Oakenpants** [comment]
I've almost closed this a dozen times, and the few times I have started to try and integrate the info, it all just gets too messy @smithfred is the first person (AFAIK) in approx 3 years, since I started this thing, to ask about cache

> I know in my case, when I came to this section, my immediate thought was "Wait, why is cache being disabled? Is this just an "on-disk browsing info available locally" problem? Is it a tracking issue (ETags?) Is there some other attack vector I don't know about they're preventing here?"

It's EXACTLY what you thought it was ("on-disk browsing info available locally" - no need to explain it, you got it in one. It's THAT obvious.

IF a pref is there for a specific threat, we explain it, but for generalized threats, there is no need.

**Comment by Thorin-Oakenpants** [comment]
>  I'm neither dumb nor uninformed

Didn't mean to imply you were. You raised a valid point see fresh eyes comment below

> Presumably you're confident that

No, I'm not confident. I have blind spots due to the fact that I have been living in this shell for 3 years and am not immune to assumptions - that's why a fresh set of eyes is always welcome. The other side of the coin is that a line has to be drawn somewhere regards verbosity, assumed level of knowledge, etc.

600+ prefs - complex intertwined hot-mess of constantly changing tech ... kinda going to go out on a limb here and say its well beyond the average user - not that I'd don't TRY to make it easier

As said elsewhere, the answer to this is a relaxed branch etc where 30-40 prefs are relaxed for basically zero breakage - then the less knowledgeable can feel more confident in using it, and slowly build up

> I listed 3 different possibilities

Your **two** possibilities ("something else" does not qualify) were both correct. The first was the main reason for disk cache. ETAGs are also controlled by disk cache but require memory cache also turned off. So you already had the answers.

**Comment by crssi** [comment]
@smithfred  if you still want to enable memory cache, then follow [https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.2.4-Header-Editor](https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.2.4-Header-Editor) to get rid of ETags.

**Comment by smithfred** [comment]
> Your two possibilities ("something else" does not qualify)

3. Local/unlocked history access; tracking; attack (vulnerability exploit) vector. They all qualify as the point is to enumerate the possible classes of problem to demonstrate the ambiguity of undocumented pref settings, not to *know them exactly*. The fact that the last one is vague only strengthens the point. New vectors crop up so frequently that someone having certainty that they know all of them (and can therefore determine with certainty without documentation the purpose of a pref set) is extremely difficult.

> So you already had the answers.

In the same way that someone who doesn't know the correct answer to a multiple-choice question already "has all the answers". I.e. only in a literal, and completely useless sense :J
